internal/cluster/cluster.go

```diff
@@ -0,0 +1,264 @@
+			kernelVersionLogger.V(1).Info("Using cached image", "image", image)
+			continue
+		}
+
+		m, err := c.kernelAPI.FindMappingForKernel(modKernelMappings, kernelVersion)
+		if err != nil {
+			kernelVersionLogger.Info("no suitable container image found; skipping kernel version")
+			continue
+		}
+
+		m, err = c.kernelAPI.PrepareKernelMapping(m, osConfig)
+		if err != nil {
+			kernelVersionLogger.Info("failed to substitute the template variables in the mapping", "error", err)
+			continue
+		}
+
+		kernelVersionLogger.V(1).Info("Found a valid mapping",
+			"image", m.ContainerImage,
+			"build", m.Build != nil,
+		)
+
+		mappings[kernelVersion] = m
+	}
+
+	return mappings, nil
+}
+
+func (c *clusterAPI) kernelVersions(cluster clusterv1.ManagedCluster) ([]string, error) {
+	for _, clusterClaim := range cluster.Status.ClusterClaims {
+		if clusterClaim.Name != clusterClaimName {
+			continue
+		}
+		return strings.Split(clusterClaim.Value, "\n"), nil
+	}
+	return nil, errors.New("KMM ClusterClaim not found")
+}
+
+func (c *clusterAPI) build(
+	ctx context.Context,
+	mod kmmv1beta1.Module,
+	mcm *kmmv1beta1.ManagedClusterModule,
+	kernelMapping *kmmv1beta1.KernelMapping,
+	kernelVersion string) (bool, error) {
+
+	shouldSync, err := c.buildAPI.ShouldSync(ctx, mod, *kernelMapping)
+	if err != nil {
+		return false, fmt.Errorf("could not check if build synchronization is needed: %v", err)
+	}
+	if !shouldSync {
+		return false, nil
+	}
+
+	logger := log.FromContext(ctx).WithValues(
+		"kernel version", kernelVersion,
+		"image", kernelMapping.ContainerImage)
+	buildCtx := log.IntoContext(ctx, logger)
+
+	buildRes, err := c.buildAPI.Sync(buildCtx, mod, *kernelMapping, kernelVersion, true, mcm)
+	if err != nil {
+		return false, fmt.Errorf("could not synchronize the build: %w", err)
+	}
+
+	return buildRes.Requeue, nil
+}
+
+func (c *clusterAPI) sign(
+	ctx context.Context,
+	mod kmmv1beta1.Module,
+	mcm *kmmv1beta1.ManagedClusterModule,
+	kernelMapping *kmmv1beta1.KernelMapping,
+	kernelVersion string) (bool, error) {
+
+	shouldSync, err := c.signAPI.ShouldSync(ctx, mod, *kernelMapping)
+	if err != nil {
+		return false, fmt.Errorf("could not check if signing synchronization is needed: %v", err)
+	}
+	if !shouldSync {
+		return false, nil
+	}
+
+	// if we need to sign AND we've built, then we must have built
+	// the intermediate image so must figure out its name
+	previousImage := ""
+	if module.ShouldBeBuilt(mcm.Spec.ModuleSpec, *kernelMapping) {
+		previousImage = module.IntermediateImageName(mod.Name, mod.Namespace, kernelMapping.ContainerImage)
+	}
+
+	logger := log.FromContext(ctx).WithValues(
+		"kernel version", kernelVersion,
+		"image", kernelMapping.ContainerImage)
+	signCtx := log.IntoContext(ctx, logger)
+
+	signRes, err := c.signAPI.Sync(signCtx, mod, *kernelMapping, kernelVersion, previousImage, true, mcm)
+	if err != nil {
+		return false, fmt.Errorf("could not synchronize the signing: %w", err)
+	}
```

Comment:
This logic is also used in `controllers/module_reconciler.go`. Can we extract that functionality to a separate API (utils for example) and reuse it instead of duplicating it?

Reply:
Same as [here](https://github.com/kubernetes-sigs/kernel-module-management/pull/178#discussion_r1042063545). Let's keep the conversation there, everything that applies there, applies here as well by default :)